File: sdks/flutter/lib/likerts.dart
import 'dart:convert';
import 'dart:async';
import 'package:http/http.dart' as http;

const likertsSdkCapability={'target':'flutter','sdkVersion':'0.0.3','schemaVersions':[1,2,3,4,5]};

class VisibilityCondition {
  final String questionId, operator;
  final dynamic value;
  VisibilityCondition.fromJson(Map<String,dynamic> json):questionId=json['questionId'],operator=json['operator'],value=json['value'];
}
class PromptItem{final String id,label;PromptItem.fromJson(Map<String,dynamic> value):id=value['id'],label=value['label'];}
class PageBranch {final VisibilityCondition when;final String goToPageId;PageBranch.fromJson(Map<String,dynamic> value):when=VisibilityCondition.fromJson(Map<String,dynamic>.from(value['when'])),goToPageId=value['goToPageId'];}
class SurveyPage {final String id;final String? title;final List<String> questionIds;final List<PageBranch> branches;SurveyPage.fromJson(Map<String,dynamic> value):id=value['id'],title=value['title'],questionIds=List<String>.from(value['questionIds']),branches=(value['branches'] as List? ?? const []).map((item)=>PageBranch.fromJson(Map<String,dynamic>.from(item))).toList();}

class Question {
  final String id, type, label;
  final bool required;
  final List<Map<String, dynamic>> options;
  final num? min, max;
  final int? maxLength, minSelections, maxSelections;
  final String? preset;
  final Map<String, String>? labels;
  final VisibilityCondition? visibleWhen;
  final String? presentation;
  final List<PromptItem> rows,items;
  final List<Map<String,dynamic>> columns;
  final String? matrixMode;
  final int? total;
  List<int> get scaleValues => type == "scale" && (preset == "nps" || labels != null || presentation == "stars") && min != null && max != null && min! >= -10000 && max! <= 10000 && min! <= max! ? List.generate(max!.toInt() - min!.toInt() + 1, (i) => min!.toInt() + i) : [];
  String scaleLabel(int value) => labels?["$value"] == null ? "$value" : "$value — ${labels!["$value"]}";
  String? selectionError(List? selected) {
    if(type != "multiple_choice") return null;
    if(selected == null) return required ? "Choose at least one option" : null;
    final minimum = (minSelections ?? 0) < (required ? 1 : 0) ? 1 : (minSelections ?? 0);
    if(selected.length < minimum) return "Choose at least $minimum options";
    if(maxSelections != null && selected.length > maxSelections!) return "Choose at most $maxSelections options";
    return null;
  }
  Question.fromJson(Map<String, dynamic> value)
      : id = value['id'], type = value['type'], label = value['label'], required = value['required'] ?? false,
        options = (value['options'] as List? ?? []).map((o) => Map<String, dynamic>.from(o)).toList(),
        min = value['min'], max = value['max'], maxLength = value['maxLength'], preset = value['preset'], labels = value['labels'] == null ? null : Map<String, String>.from(value['labels']), minSelections = value['minSelections'], maxSelections = value['maxSelections'], visibleWhen=value['visibleWhen']==null?null:VisibilityCondition.fromJson(Map<String,dynamic>.from(value['visibleWhen'])),presentation=value['presentation'],rows=(value['rows'] as List? ?? []).map((v)=>PromptItem.fromJson(Map<String,dynamic>.from(v))).toList(),columns=(value['columns'] as List? ?? []).map((v)=>Map<String,dynamic>.from(v)).toList(),matrixMode=value['matrixMode'],items=(value['items'] as List? ?? []).map((v)=>PromptItem.fromJson(Map<String,dynamic>.from(v))).toList(),total=value['total'];
}
class Collection {
  final String id, surveyId, placement, title;
  final int version;
  final int schemaVersion;
  final List<Question> questions;
  final List<SurveyPage>? pages;
  Collection.fromJson(Map<String, dynamic> value)
      : id = value['id'], surveyId = value['surveyId'], placement = value['placement'], version = value['version'], schemaVersion=value['schema']['schemaVersion'],
        title = value['schema']['title'], questions = (value['schema']['questions'] as List).map((q) => Question.fromJson(q)).toList(), pages=value['schema']['pages']==null?null:(value['schema']['pages'] as List).map((page)=>SurveyPage.fromJson(Map<String,dynamic>.from(page))).toList() {
    if (![1, 2, 3, 4, 5].contains(value['schema']['schemaVersion'])) throw const FormatException('Unsupported schema');
  }
}
/// Generate and persist a unique key in the host app; reuse this immutable payload on retries.
class Submission {
  final String idempotencyKey;
  final Map<String, dynamic> answers, metadata;
  Submission({required this.idempotencyKey, required Map<String, dynamic> answers, Map<String, dynamic> metadata = const {}})
      : answers = Map.unmodifiable(answers), metadata = Map.unmodifiable(metadata);
  Map<String, dynamic> toJson() => {'idempotencyKey': idempotencyKey, 'answers': answers, 'metadata': metadata};
}
class LikertsException implements Exception {
  final int status;
  final String response;
  LikertsException(this.status, this.response);
  @override String toString() => 'Likerts request failed ($status)';
}
class Receipt {
  final String responseId, collectionId;
  final bool accepted;
  final int chargedCents;
  Receipt.fromJson(Map<String,dynamic> value)
      : responseId=value['responseId'], collectionId=value['collectionId'], accepted=value['accepted'], chargedCents=value['chargedCents'] {
    if(!accepted || chargedCents != 1) throw const FormatException('Invalid Likerts receipt');
  }
}
class LikertsCancellationToken {
  final Completer<void> _cancelled=Completer<void>();
  Future<void> get whenCancelled => _cancelled.future;
  bool get isCancelled => _cancelled.isCompleted;
  void cancel() { if(!_cancelled.isCompleted) _cancelled.complete(); }
}
class LikertsClient {
  final String baseUrl, collectionToken;
  final Duration timeout;
  final Duration cacheMaxAge;
  final http.Client _http;
  final Map<String,({Collection value,DateTime storedAt})> _collectionCache={};
  LikertsClient({required String baseUrl, required this.collectionToken, http.Client? client, this.timeout=const Duration(seconds:15),this.cacheMaxAge=const Duration(minutes:5)}) : baseUrl=_safeBaseUrl(baseUrl), _http = client ?? http.Client() {
    if(timeout <= Duration.zero) throw ArgumentError.value(timeout,'timeout','must be positive');
    if(cacheMaxAge.isNegative) throw ArgumentError.value(cacheMaxAge,'cacheMaxAge','must not be negative');
  }
  static String _safeBaseUrl(String value) {
    final uri=Uri.parse(value);final loopback=const {'localhost','127.0.0.1','::1'}.contains(uri.host.toLowerCase());
    if(!(uri.scheme == 'https' || (uri.scheme == 'http' && loopback)) || uri.userInfo.isNotEmpty || uri.hasQuery || uri.hasFragment) throw ArgumentError.value(value,'baseUrl','Likerts base URL must use HTTPS (HTTP is limited to loopback development)');
    return value.replaceFirst(RegExp(r"/$"), "");
  }
  Future<Map<String, dynamic>> _request(String id, [Submission? submission, LikertsCancellationToken? cancellation]) async {
    final uri = Uri.parse('${baseUrl.replaceFirst(RegExp(r"/$"), "")}/v1/collections/${Uri.encodeComponent(id)}${submission == null ? "" : "/responses"}');
    final timeoutAbort=Completer<void>();
    final abort=Future.any<void>([timeoutAbort.future,if(cancellation != null)cancellation.whenCancelled]);
    final request=http.AbortableRequest(submission == null ? 'GET' : 'POST',uri,abortTrigger:abort)
      ..followRedirects=false
      ..headers.addAll({'Authorization':'Bearer $collectionToken','Content-Type':'application/json'});
    if(submission != null) request.body=jsonEncode(submission.toJson());
    final send=_http.send(request);
    final cancelled=abort.then<http.StreamedResponse>((_)=>throw http.RequestAbortedException(uri));
    late http.StreamedResponse streamed;
    try { streamed=await Future.any([send,cancelled]).timeout(timeout); }
    on TimeoutException { if(!timeoutAbort.isCompleted)timeoutAbort.complete();rethrow; }
    final response=await http.Response.fromStream(streamed);
    if (response.statusCode < 200 || response.statusCode >= 300) throw LikertsException(response.statusCode, response.body);
    return jsonDecode(response.body) as Map<String, dynamic>;
  }
  Future<Collection> collection(String id,{LikertsCancellationToken? cancellation,bool refresh=false}) async {
    final cached=_collectionCache[id];if(!refresh && cached != null && DateTime.now().difference(cached.storedAt)<cacheMaxAge)return cached.value;
    try { final value=Collection.fromJson(await _request(id,null,cancellation));if(cached != null && (cached.value.id != value.id || cached.value.surveyId != value.surveyId || cached.value.version != value.version || cached.value.schemaVersion != value.schemaVersion)){_collectionCache.remove(id);throw const FormatException('Collection binding changed');}_collectionCache[id]=(value:value,storedAt:DateTime.now());return value; }
    catch(error){_collectionCache.remove(id);rethrow;}
  }
  void clearCollectionCache([String? id]) { if(id == null){_collectionCache.clear();}else{_collectionCache.remove(id);} }
  Future<Receipt> submit(String id, Submission submission,{LikertsCancellationToken? cancellation}) async => Receipt.fromJson(await _request(id,submission,cancellation));
  void close() => _http.close();
}

File: sdks/android/src/main/kotlin/com/likerts/sdk/Likerts.kt
package com.likerts.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runInterruptible
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import java.net.HttpURLConnection
import java.net.URI
import java.net.URLEncoder
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

@Serializable data class SDKInstallationCapability(val target: String, val sdkVersion: String, val schemaVersions: List<Int>)
val LIKERTS_SDK_CAPABILITY = SDKInstallationCapability("android", "0.0.3", listOf(1, 2, 3, 4, 5))

@Serializable data class OtherText(val maxLength:Int)
@Serializable data class Option(val id: String, val label: String, val other:OtherText?=null,val exclusive:Boolean?=null)
@Serializable data class VisibilityCondition(val questionId:String,val operator:String,val value:JsonPrimitive?=null)
@Serializable data class PromptItem(val id:String,val label:String)
@Serializable data class PageBranch(val `when`:VisibilityCondition,val goToPageId:String)
@Serializable data class SurveyPage(val id:String,val title:String?=null,val questionIds:List<String>,val branches:List<PageBranch> = emptyList())
@Serializable data class Question(val id: String, val type: String, val label: String, val required: Boolean = false, val options: List<Option> = emptyList(), val min: Double? = null, val max: Double? = null, val maxLength: Int? = null, val preset: String? = null, val labels: Map<String, String>? = null, val minSelections: Int? = null, val maxSelections: Int? = null,val visibleWhen:VisibilityCondition?=null,val presentation:String?=null,val rows:List<PromptItem>?=null,val columns:List<Option>?=null,val matrixMode:String?=null,val items:List<PromptItem>?=null,val total:Int?=null) {
    fun scaleValues(): List<Int> = if (type == "scale" && (preset == "nps" || labels != null || presentation == "stars") && min != null && max != null && min >= -10000 && max <= 10000 && min <= max) (min.toInt()..max.toInt()).toList() else emptyList()
    fun scaleLabel(value: Int): String = labels?.get(value.toString())?.let { "$value — $it" } ?: value.toString()
    fun selectionError(selected: kotlin.collections.Collection<String>?): String? {
        if (type != "multiple_choice") return null
        if (selected == null) return if (required) "Choose at least one option" else null
        val minimum = maxOf(minSelections ?: 0, if (required) 1 else 0)
        return when { selected.size < minimum -> "Choose at least $minimum options"; maxSelections != null && selected.size > maxSelections -> "Choose at most $maxSelections options"; else -> null }
    }
    fun validationError(textValue: String?, selected: kotlin.collections.Collection<String>?): String? {
        val missing = if(type == "multiple_choice") selected.isNullOrEmpty() else textValue.isNullOrEmpty()
        if(required && missing) return "$label: an answer is required."
        return selectionError(selected)
    }
}
@Serializable data class SurveySchema(val schemaVersion: Int, val title: String, val questions: List<Question>,val pages:List<SurveyPage>?=null) { init { require(schemaVersion in 1..5) { "Unsupported schema" } } }
@Serializable data class Collection(val id: String, val surveyId: String, val version: Int, val placement: String, val schema: SurveySchema)
@Serializable data class Submission(val idempotencyKey: String = UUID.randomUUID().toString(), val answers: Map<String, JsonElement>, val metadata: JsonObject = JsonObject(emptyMap()))
@Serializable data class Receipt(val responseId: String, val collectionId: String, val accepted: Boolean) { init { require(accepted) { "Invalid Likerts receipt" } } }
class LikertsException(val status: Int, val response: String): Exception("Likerts request failed ($status)")
/** Collection-only credential. Reuse the same Submission on transport retries. */
class LikertsClient(private val baseUrl: String, private val collectionToken: String, private val connectTimeoutMillis: Int = 15000, private val readTimeoutMillis: Int = 15000, private val cacheMaxAgeMillis: Long = 300000) {
    private val json = Json { ignoreUnknownKeys = true }
    private data class Cached(val value: Collection, val storedAtNanos: Long)
    private val collectionCache = ConcurrentHashMap<String, Cached>()
    private fun baseUri(): URI {
        val uri=URI.create(baseUrl)
        val loopback=uri.host?.lowercase() in setOf("localhost","127.0.0.1","::1")
        require((uri.scheme == "https" || (uri.scheme == "http" && loopback)) && uri.userInfo == null && uri.query == null && uri.fragment == null) { "Likerts base URL must use HTTPS (HTTP is limited to loopback development)" }
        require(connectTimeoutMillis > 0 && readTimeoutMillis > 0) { "Likerts timeouts must be positive" }
        require(cacheMaxAgeMillis >= 0) { "Likerts cache max age must not be negative" }
        return uri
    }
    private suspend fun request(id: String, submission: Submission? = null): String = runInterruptible(Dispatchers.IO) {
        val segment = URLEncoder.encode(id, "UTF-8").replace("+", "%20")
        val url = URI.create("${baseUri().toString().trimEnd('/')}/v1/collections/$segment" + if (submission != null) "/responses" else "").toURL()
        val connection = url.openConnection() as HttpURLConnection
        try {
            connection.connectTimeout = connectTimeoutMillis; connection.readTimeout = readTimeoutMillis
            connection.instanceFollowRedirects = false
            connection.setRequestProperty("Authorization", "Bearer $collectionToken")
            if (submission != null) { connection.requestMethod = "POST"; connection.doOutput = true; connection.setRequestProperty("Content-Type", "application/json"); connection.outputStream.use { it.write(json.encodeToString(submission).toByteArray(Charsets.UTF_8)) } }
            val code = connection.responseCode
            val response = (if (code in 200..299) connection.inputStream else connection.errorStream)?.bufferedReader()?.use { it.readText() } ?: ""
            if (code !in 200..299) throw LikertsException(code, response)
            response
        } finally { connection.disconnect() }
    }
    suspend fun collection(id: String, refresh: Boolean = false): Collection {
        val cached=collectionCache[id]
        if(!refresh && cached != null && (System.nanoTime()-cached.storedAtNanos)/1_000_000 < cacheMaxAgeMillis) return cached.value
        try { val result=json.decodeFromString<Collection>(request(id));require(result.schema.schemaVersion in 1..5){"Unsupported schema"}
            if(cached != null && (cached.value.id != result.id || cached.value.surveyId != result.surveyId || cached.value.version != result.version || cached.value.schema.schemaVersion != result.schema.schemaVersion)){collectionCache.remove(id);throw IllegalStateException("Collection binding changed")}
            collectionCache[id]=Cached(result,System.nanoTime());return result
        } catch(error: Throwable) { collectionCache.remove(id);throw error }
    }
    fun clearCollectionCache(id: String? = null) { if(id == null) collectionCache.clear() else collectionCache.remove(id) }
    suspend fun submit(id: String, submission: Submission): Receipt = json.decodeFromString(request(id, submission))
}
